gramineproject/gramine
# Python test runner

## Description of the changes 

This PR converts building test manifest from Makefiles to our own Python wrapper, `gramine-test`. See #161 for discussion.

## Why Ninja?

The wrapper uses **Ninja** internally, i.e. generates `build.ninja` and runs Ninja to build manifests. I believe I have good reasons for prefering Ninja over a completely custom Python script:

* Ninja handles concurrent building
* Ninja tracks dependencies, i.e. rebuilds only what is necessary
* It's very fast! (see below)
* Most importantly (IMO), it forces us to split the layers, i.e. invoke `gramine-manifest` instead of being tempted to call a Python function.

On the opposite end of the spectrum, there's implementing everything in Meson, but I think would not be much simpler (the part that generates Ninja would translate to `custom_target`s in Meson), and the end result would be less friendly for the user.

## LTP

LTP still uses the Makefiles. My plan for LTP is to treat is the same as other tests: move building the binaries to Meson, and building manifests/running to `gramine-tests`.

When that's done, we'll be able to (finally!) remove all these old scripts: `Makefile.Test`, `manifest.mk`, etc.

## Benchmark

I did a test on an 8-core machine, in `LibOS/shim/test/regression`.

`master` (3db1e28e27622cbf1265650cc92a87281803d31d):

```
$ make clean; time make -j8
real	0m17.696s
user	0m32.239s
sys	0m3.517s

$ make SGX=1 clean; time make -j8 SGX=1 all sgx-tokens
real	0m44.426s
user	3m50.492s
sys	0m16.938s
```

This PR:

```
$ gramine-test clean; time gramine-test build
real	0m4.117s
user	0m28.205s
sys	0m2.719s

$ gramine-test --sgx clean; time gramine-test --sgx build
real	0m26.453s
user	3m13.935s
sys	0m13.083s
```

Also, reducing the static PAL buffer (#192) makes that even faster:
```
$ gramine-test --sgx clean; time gramine-test --sgx build
real	0m18.653s
user	2m12.166s
sys	0m13.671s
```

## How to test this PR? 

See how it works for you. Try using the `gramine-test` utility.


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gramineproject/gramine/182)
